src/hotspot/share/services/memReporter.cpp

```diff
@@ -820,3 +822,68 @@ void MemDetailDiffReporter::diff_virtual_memory_site(const NativeCallStack* stac
 
   out->print_cr(")\n");
  }
+
+void MemJFRReporter::sendTotalEvent() {
+  if (!MemTracker::enabled()) {
+    return;
+  }
+
+  MemBaseline usage;
+  usage.baseline(true);
```

Comment:
In an ideal world we could report totals and category information from the same snapshot, however the periodic event infrastructure doesn’t seem to provide a great way to do that. I don’t think that’s a big deal, but may be worth documenting to avoid confusion as totals may be reported lower than the sum of the parts due to temporal drift.

An option I had considered is to define some form of larger “NMTSummary” event with reserved+committed fields for each category as well as the totals. The reporting structure would align with the existing NMT summary tooling, but the event would need to be kept up to date with new NMT types, which may not be worthwhile (unless we can create dynamic events using the c apis, which would make things easier). Unclear how this data would be visualized out of the box. I don’t have as much context as I’d like around event design considerations, there may be other reasons I’m not aware of that this suggestion is a bad idea.

Reply:
I think we should design the events as if we could take everything in one snapshot. That way, we can improve the implementation later. It would be good to write in the event description of NativeMemoryUsage event that it may not add exactly up to the sum of NativeMemoryUsagePart events due to timing issues.

I will look into adding support for snapshots, but it will have to be a later release.

Reply:
Today, the call into JVM has the signature JVM::emitEvent(long eventTypeId, long timestamp, long when). If the timestamp is stored in jfr_periodic_timestamp when it first gets into native and the part event is emitted first, then it could look something like this:

static u8 jfr_periodic_timestamp;

struct JfrNmtTotal {
  u8 timestamp;
  u8 reserved;
  u8 committed;
};

static JfrNmtTotal last_nmt;

TRACE_REQUEST_FUNC(NativeMemoryUsagePart) {
   ...
   last_nmt.timestamp = jfr_periodic_timestamp;
   last_nmt.reserved = sum of reserved;
   last_nmt.committed = sum of committed;
}

TRACE_REQUEST_FUNC(NativeMemoryUsage) {
   if (jfr_periodic_timestamp == last_nmt.timestamp) {
       // use last_nmt
   } else {
      // calculate reserved and committed independently
  }
}

Periodic events are run in the same thread.

I don't know if the order of periodic event is the same as in metadata.xml but it could be fixed.

Reply:
I just realized there is another problem with trying to share the baseline, the events might be configured to have different periods and then sharing will be strange. So step one: document it.